tools/pubsys/src/repo/validate_repo/mod.rs

```diff
@@ -0,0 +1,186 @@
+//! The validate_repo module owns the 'validate-ami' subcommand and provides methods for validating
+//! a given set of TUF repositories by downloading its targets and check for metadata expirations.
+
+use super::RepoTransport;
+use crate::config::InfraConfig;
+use crate::repo::{compose_repo_urls, error as repo_error};
+use crate::Args;
+use log::{debug, error, info, trace};
+use rand::seq::SliceRandom;
+use snafu::{OptionExt, ResultExt};
+use std::fs::File;
+use std::io;
+use structopt::StructOpt;
+use tempfile::tempdir;
+use tough::{ExpirationEnforcement, Limits, Repository, Settings};
+use url::Url;
+
+type Result<T> = std::result::Result<T, crate::repo::error::Error>;
+
+/// Validates a set of TUF repositories
+#[derive(Debug, StructOpt)]
+#[structopt(setting = clap::AppSettings::DeriveDisplayOrder)]
+pub(crate) struct ValidateRepoArgs {
+    #[structopt(long)]
+    /// Use this named set of repositories from Infra.toml
+    repo: String,
+
+    #[structopt(default_value = "100", long)]
+    /// Randomly samples and retrieves specified percentage of targets
+    percent_target_files: u8,
+}
+
+/// Randomly samples specified percentage of listed targets in the TUF repo and tries to retrieve them
+fn retrieve_percentage_of_targets<T>(
+    repo: &Repository<'_, T>,
+    percentage: u8,
+) -> Option<repo_error::Error>
+where
+    T: tough::Transport,
+{
+    let targets = &repo.targets().signed.targets;
+    let percentage = f32::from(percentage) / 100.0;
+    let num_to_retrieve = (targets.len() as f32 * percentage).ceil();
+    let mut rng = &mut rand::thread_rng();
+    let mut sampled_targets: Vec<String> = targets.keys().map(|key| key.to_string()).collect();
+    sampled_targets = sampled_targets
+        .choose_multiple(&mut rng, num_to_retrieve as usize)
+        .cloned()
+        .collect();
+    for target in sampled_targets {
+        let max_attempts = 5;
+        for attempt in 1..=max_attempts {
+            if attempt != 1 {
+                debug!(
+                    "Starting attempt {} of {} for target {}",
+                    attempt, max_attempts, target
+                );
+            }
+            let target_reader = repo.read_target(&target);
+            match target_reader {
+                Err(err) => match err {
+                    tough::error::Error::Transport { .. } => {
+                        error!("Fetch failure on attempt {}", attempt);
+                        if attempt >= max_attempts {
+                            return Some(repo_error::Error::ReadTarget {
+                                target,
+                                source: err,
+                            });
```

Comment:
The logic from here on down is fairly nested - I'm wondering if it should get split out into a function so it's easier to follow and read.

Reply:
I feel like it would be somewhat more confusing if I split out the stuff in `match target_reader` into it's own function due to having to keep track of the current iteration of the loop for the purpose of implementing retries and determining the proper error to return. 

I can add more comments for clarity. Do you think that'll help?

Reply:
Scratch the above, I think I know what I can do.